be/src/io/writable.h

```diff
@@ -0,0 +1,36 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include "common/status.h"
+
+namespace starrocks::io {
+
+class Writable {
+public:
+    virtual ~Writable() = default;
+
+    // Write the given data to the stream
+    //
+    // This method always processes the bytes in full. Depending on the
+    // semantics of the stream, the data may be written out immediately,
+    // held in a buffer, or written asynchronously. In the case where
+    // the stream buffers the data, it will be copied.
+    virtual Status write(const void* data, int64_t size) = 0;
+
+    virtual bool allows_aliasing() const = 0;
+
+    // Write a given chunk of data to the output.
+    // Some output streams may implement this in a way that avoids copying.
+    // Check allows_aliasing() before calling write_aliased(). It will fall
+    // back to copying if write_aliased() is called on a stream that does not
+    // allow aliasing.
+    // NOTE: It is caller's responsibility to ensure that the chunk of memory
+    // remains live until all of the data has been consumed from the stream.
+    virtual Status write_aliased(const void* data, int64_t size) = 0;
```

Comment:
what's the different between write_aliased and write. And why its name is write_aliased.
Can you give a example to help me to understand how this can be achieved.

Reply:
> 

The main difference between `write(const void*data, int64_t size)` and `write_aliased(const void*data, int64_t size)` is the memory ownership of data, the memory passed to `write` is owned by the caller while the memory passed to `write_aliased` is owned by the `Writable`.
Suppose that we have a synchronized Writable implementation,  the chunk of data passed to `write_aliased` can be passed to the background writing thread directly without copying, while the chunk of data passed to `write` has to be copied to the internal buffer.
The name `write_aliased` is modeled from the `WriteAliasedRaw` method ZeroCopyOutputStream in [protobuf](https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.io.zero_copy_stream#ZeroCopyOutputStream.WriteAliasedRaw.details)

Reply:
> 

Supplement: it's inaccurate to say that `the memory passed to write_aliased is owned by the Writable`, the memory passed to write_aliased is still owned by the caller, but it's a calling convention that the memory will not be released or modified until the end of the lifecycle of `Writable` object.

Reply:
I see, Thanks for your explanation. Could you mind add this explanation as comment.

Reply:
updated, PTAL